### ISSUE TITLE
expo-haptics crashes

### ISSUE THREAD

**Issue by yhkaplan** [expo/expo#19127]
### Summary

UIFeedbackGenerator APIs are not thread-safe and should only be called on the main thread, or else the deactivated calls occurring on a thread other than the initial caller appear to cause crashes. The current Swift implementation inside an "AsyncFunction" should either hop back to the main thread or be changed to be a synchronous functions on the main thread. I made a patch in my codebase that called generator.prepare() and generator.impactOccurred() etc on the main thread, and the crashes below entirely disappeared. 

I can make a PR, but what solution is preferred: wrapping the calls in DispatchQueue.main.async (which would involve some extraneous thread-hopping) or a larger API change to synchronous functions?

As for reproducibility, the crashes occur when we have thousands of users, so it is not easily reproduced locally. That said, the current Apple documentation page shows how UIFeedbackGenerator is denoted as `@MainActor` in Swift Concurrency compatible versions of Swift, proving that method calls should occur on the main thread as well. 

https://developer.apple.com/documentation/uikit/uifeedbackgenerator

```
Crashed: com.apple.main-thread
0  libdispatch.dylib              0x246c _os_object_release + 76
1  UIKitCore                      0x3795c4 -[UIFeedbackGenerator _deactivated] + 40
2  UIKitCore                      0xa6c9bc -[UISelectionFeedbackGenerator _deactivated] + 52
3  UIKitCore                      0x578114 -[UIFeedbackGenerator __deactivateWithStyle:] + 908
4  libdispatch.dylib              0x1e6c _dispatch_call_block_and_release + 32
5  libdispatch.dylib              0x3a30 _dispatch_client_callout + 20
6  libdispatch.dylib              0x11f48 _dispatch_main_queue_drain + 928
7  libdispatch.dylib              0x11b98 _dispatch_main_queue_callback_4CF + 44
8  CoreFoundation                 0x51800 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
9  CoreFoundation                 0xb704 __CFRunLoopRun + 2532
10 CoreFoundation                 0x1ebc8 CFRunLoopRunSpecific + 600
11 GraphicsServices               0x1374 GSEventRunModal + 164
12 UIKitCore                      0x514b58 -[UIApplication _run] + 1100
13 UIKitCore                      0x296090 UIApplicationMain + 364
14 $APP_NAME                        0x8554 main + 8 (main.m:8)
15 ???                            0x101d6dda4 (Missing)
```

```
Crashed: com.apple.root.default-qos
0  libdispatch.dylib              0x246c _os_object_release + 76
1  UIKitCore                      0x379150 -[UIFeedbackGenerator _deactivated] + 40
2  UIKitCore                      0xa6bf7c -[UISelectionFeedbackGenerator _deactivated] + 52
3  UIKitCore                      0x577c48 -[UIFeedbackGenerator __deactivateWithStyle:] + 908
4  UIKitCore                      0x485d74 -[UIFeedbackGenerator _deactivateWithStyle:] + 344
5  UIKitCore                      0x3f6bf0 -[UIFeedbackGenerator _stopPreparationForStyle:] + 76
6  UIKitCore                      0x69198c -[UIFeedbackGenerator _playFeedback:withMinimumIntervalPassed:since:prefersRegularPace:] + 452
7  UIKitCore                      0x67d6d0 -[UISelectionFeedbackGenerator _playFeedbackWithSpeed:] + 708
8  $APP_NAME                        0x16feb8 specialized thunk for @escaping @callee_guaranteed () -> (@error @owned Error) + 4367662776 (<compiler-generated>:4367662776)
9  $APP_NAME                        0x192fe8 thunk for @escaping @callee_guaranteed () -> (@out A, @error @owned Error) + 4367806440 (<compiler-generated>:4367806440)
10 $APP_NAME                        0x1b830c partial apply for thunk for @escaping @callee_guaranteed () -> (@out A, @error @owned Error) + 4367958796 (<compiler-generated>:4367958796)
11 $APP_NAME                        0x199e9c ConcreteFunction.call(args:promise:) + 54 (ConcreteFunction.swift:54)
12 $APP_NAME                        0x19acc0 protocol witness for AnyFunction.call(args:promise:) in conformance ConcreteFunction<A, B> + 4367838400 (<compiler-generated>:4367838400)
13 $APP_NAME                        0x1b56f4 partial apply for closure #1 in ModuleHolder.call(function:args:promise:) + 4367947508 (<compiler-generated>:4367947508)
14 $APP_NAME                        0x1772d8 thunk for @escaping @callee_guaranteed () -> () + 4367692504 (<compiler-generated>:4367692504)
15 libdispatch.dylib              0x1e6c _dispatch_call_block_and_release + 32
16 libdispatch.dylib              0x3a30 _dispatch_client_callout + 20
17 libdispatch.dylib              0x6b44 _dispatch_queue_override_invoke + 784
18 libdispatch.dylib              0x15164 _dispatch_root_queue_drain + 396
19 libdispatch.dylib              0x1596c _dispatch_worker_thread2 + 164
20 libsystem_pthread.dylib        0x1080 _pthread_wqthread + 228
21 libsystem_pthread.dylib        0xe5c start_wqthread + 8
```

### What platform(s) does this occur on?

iOS

### Environment

Expo Haptics versions written in Swift

### Minimal reproducible example

Basic calls to below functions on iOS devices

```
Haptics.selectionAsync()
Haptics.impactAsync(...)
Haptics.notificationAsync(...)
```

**Comment by AntonGolikov** [comment]
Hey, it looks like my PR #19819 addresses your issue as well. Should be solved now

**Comment by brentvatne** [comment]
fixed by @AntonGolikov in #19819 and released in sdk 47!
